src/relayclient/KnownRelaysManager.ts

```diff
@@ -0,0 +1,146 @@
+  }
+
+  /**
+   * Basic score is transaction fee(%), higher is better.
+   * Relays that failed to respond recently will be downgraded for some period of time.
+   */
+  calculateRelayScore (relay: RelayRegisteredEventInfo): number {
+    let score = 1000 - parseInt(relay.pctRelayFee)
+
+    const latestRelayFailure = this.latestRelayFailures.get(relay.relayUrl)
+    if (latestRelayFailure != null) {
+      const elapsed = (new Date().getTime() - latestRelayFailure.lastErrorTime) / 1000
+      // relay failed to answer lately and it's score will be downgraded
+      const timeoutGrace = this.config.relayTimeoutGrace ?? DEFAULT_RELAY_TIMEOUT_GRACE_SEC
+      if (elapsed < timeoutGrace) {
+        score -= 10
+      } else {
+        this.latestRelayFailures.delete(relay.relayUrl)
+      }
+    }
+    return score
+  }
+
+  /**
+   * Iterates through all relevant logs emitted by GSN contracts
+   * initializes an array {@link activeRelays}
+   */
+  async refresh (): Promise<void> {
+    const relayHub = await this.contractInteractor._createRelayHub(this.relayHubAddress)
+    const relayManagers = await this._fetchRecentlyActiveRelayManagers(relayHub)
+    const relayServerRegisteredTopic = event2topic(relayHub, 'RelayServerRegistered')
+
+    // found all addresses. 2nd round to get the RelayAdded event for each of those relays.
+    // TODO: at least some of the found relays above was due to "RelayAdded" event,
+    // we _could_ optimize for that, but since at least _some_ relays
+    // were found by the TransactionRelayed event, we are forced to search them
+    // for actual address.
+    const relayServerRegisteredEvents: any[] = await relayHub.contract.getPastEvents('RelayServerRegistered', {
+      fromBlock: 1,
+      topics: [relayServerRegisteredTopic,
+        Array.from(relayManagers,
+          (address: Address) => `0x${address.replace(/^0x/, '').padStart(64, '0').toLowerCase()}`
+        )]
+    })
+
+    if (this.config.verbose) {
+      console.log(`== fetchRelaysAdded: found ${relayServerRegisteredEvents.length} unique RelayAdded events (should have at least as unique relays, above)`)
+    }
+
+    const activeRelays = new Map<Address, RelayRegisteredEventInfo>()
+    relayServerRegisteredEvents.forEach(event => {
+      const args = event.returnValues
+      const relay = {
+        relayManager: args.relayManager,
+        relayUrl: args.url,
+        baseRelayFee: args.baseRelayFee,
+        pctRelayFee: args.pctRelayFee
+      }
+      // relay.score = this.calculateRelayScore(relay)
+      activeRelays.set(args.relayManager, relay)
+    })
+    const origRelays = Object.values(activeRelays)
+    const filteredRelays = origRelays.filter(this.relayFilter)
+    filteredRelays.forEach(relay => this.activeRelays.add(relay))
+  }
+
+  async _fetchRecentlyActiveRelayManagers (relayHub: IRelayHubInstance): Promise<Set<Address>> {
+    const fromBlock = this.latestScannedBlock
```

Comment:
I am now creating tests for KnownRelaysManager. Will add support for 'stake manager' events and configurable lookup window. This method would obviously fetch all RelayHub events  it can, which probably won't work on mainnet.

Reply:
the logic we agreed upon is:
- scan latest events (from all relays) in the past "X" time
- filter unique relays (and filter-out removed ones)
- find all "RegisterRelay" of those unique relays

the time-window "X" should be large enough, so a relay doesn't have to re-register if it served any transaction by clients - otherwise, clients prefer relays who registered lately, and require them (again) to re-register periodically.